test/utilities/test_contains_none_values.py

```diff
@@ -0,0 +1,34 @@
+from profiles.utilities import contains_none_values
+
+
+def test_contains_none_values() -> None:
```

Comment:
personally I'd probably split into separate tests

Reply:
Should be parametrised, so we can see all the failures rather than just 1.

Reply:
https://docs.pytest.org/en/latest/reference.html#pytest-mark-parametrize-ref ?

Reply:
yes, parameterise is good too. Separate methods would have the advantage of better expressing what they are testing. e.g. what makes `dict_3` and `dict_4` different..

Reply:
How does adding `pytest.mark.parametrize` improve this test?

Reply:
@de-code `dict_3` does not contain a `None` value and, when passed to the function, should return `False`. `dict_4` does contain a `None` value and, when passed to the function, should return `True`. Do we need separate tests to explain this?

Reply:
@GiancarloFusiello I might be able to infer it from looking at it a bit longer but it's not so clear when the intention is not expressed. There could also be a _bug_ in the test data. Could go further, why do we need `dict_4` and `dict_5` when we have `dict_2`. I could guess what the reason was but it's getting less clear with `dict_5`. If I had to modify the behaviour in the future it wouldn't be as easy to tell whether the test is still valid. To me these are all separate test cases.

Another way to express the intention more would be to name the variables more distinctly (or use constants). e.g. `DICT_WITH_NESTED_NONE_VALUE` (or something like that)

Currently the test input data (`dict_1` etc) and expected output (`True`/`False`) seem to be a bit far apart. They would naturally sit close if they were in a separate test method but could be moved together if not.

`pytest.mark.parametrize` would also move them together and run each test individually (you would parameterise a list of _input data_ and _expected output_ pairs). It could then run all of the tests even if one was failing. I find parameterise may be more clear where the expected output is the same or similar. e.g. input values with different character sets. Or maybe where the test method is a bit more complex to avoid repetition.

Just my thoughts (maybe I just need more sleep).

Reply:
@de-code I understand now. The variable names used in the test case do not describe the value. You're right, I'll update these.

Admittedly not having used pytest much in the past, I didn't fully understand the benefits of using `pytest.mark.parametrize` by reading the documentation but, after having found some examples, I can see how this can be useful.

I'll update the code and push these changes.